pkg/cmd/cli/cmd/debug.go

```diff
@@ -0,0 +1,514 @@
+	if i := cmd.ArgsLenAtDash(); i != -1 && i < len(args) {
+		o.Command = args[i:]
+		args = args[:i]
+	}
+	switch {
+	case o.ForceTTY && o.NoStdin:
+		return kcmdutil.UsageError(cmd, "you may not specify -I and -t together")
+	case o.ForceTTY && o.DisableTTY:
+		return kcmdutil.UsageError(cmd, "you may not specify -t and -T together")
+	case o.ForceTTY:
+		o.Attach.TTY = true
+	case o.DisableTTY:
+		o.Attach.TTY = false
+	default:
+		o.Attach.TTY = term.IsTerminal(in)
+	}
+	if o.NoStdin {
+		o.Attach.TTY = false
+		o.Attach.Stdin = false
+	}
+
+	if o.Annotations == nil {
+		o.Annotations = make(map[string]string)
+	}
+
+	cmdNamespace, explicit, err := f.DefaultNamespace()
+	if err != nil {
+		return err
+	}
+
+	mapper, typer := f.Object()
+	b := resource.NewBuilder(mapper, typer, resource.ClientMapperFunc(f.ClientForMapping), kapi.Codecs.UniversalDecoder()).
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		SingleResourceType().
+		ResourceTypeOrNameArgs(true, args...).
+		Flatten()
+	if len(o.Filename) > 0 {
+		b.FilenameParam(explicit, o.Filename)
+	}
+
+	one := false
+	infos, err := b.Do().IntoSingular(&one).Infos()
+	if err != nil {
+		return err
+	}
+	if !one {
+		return fmt.Errorf("you must identify a resource with a pod template to debug")
+	}
+
+	pod, err := f.AttachablePodForObject(infos[0].Object)
+	if err != nil {
+		return fmt.Errorf("cannot debug %s: %v", infos[0].Name, err)
+	}
+	pod.Name, pod.Namespace = infos[0].Name, infos[0].Namespace
+	o.Attach.Pod = pod
+
+	if len(o.Attach.ContainerName) == 0 && len(pod.Spec.Containers) > 0 {
+		glog.V(4).Infof("defaulting container name to %s", pod.Spec.Containers[0].Name)
+		o.Attach.ContainerName = pod.Spec.Containers[0].Name
+	}
+
+	o.Annotations[debugPodAnnotationSourceResource] = fmt.Sprintf("%s/%s", infos[0].Mapping.Resource, infos[0].Name)
+	o.Annotations[debugPodAnnotationSourceContainer] = o.Attach.ContainerName
+
+	output := kcmdutil.GetFlagString(cmd, "output")
+	if len(output) != 0 {
+		o.Print = func(pod *kapi.Pod, out io.Writer) error {
+			return f.PrintObject(cmd, pod, out)
+		}
+	}
+
+	config, err := f.ClientConfig()
+	if err != nil {
+		return err
+	}
+	o.Attach.Config = config
+
+	_, kc, err := f.Clients()
+	if err != nil {
+		return err
+	}
+	o.Attach.Client = kc
+	return nil
+}
+func (o DebugOptions) Validate() error {
+	names := containerNames(o.Attach.Pod)
+	if len(names) == 0 {
+		return fmt.Errorf("the provided pod must have at least one container")
+	}
+	if len(o.Attach.ContainerName) == 0 {
+		return fmt.Errorf("you must provide a container name to debug")
+	}
+	if containerForName(o.Attach.Pod, o.Attach.ContainerName) == nil {
+		return fmt.Errorf("the container %q is not a valid container name; must be one of %v", o.Attach.ContainerName, names)
+	}
+	return nil
+}
+
+type WatchConditionFunc func(event watch.Event) (bool, error)
+
+func Until(timeout time.Duration, watch watch.Interface, conditions ...WatchConditionFunc) error {
```

Comment:
It probably makes sense regardless except for watches on lists - not sure
what to do there.

On Wed, Feb 17, 2016 at 11:00 AM, David Eads <notifications@github.com>
wrote:

> In pkg/cmd/cli/cmd/debug.go
> <https://github.com/openshift/origin/pull/7294#discussion_r53185188>:
>
> > +	names := containerNames(o.Attach.Pod)
> > +	if len(names) == 0 {
> > +		return fmt.Errorf("the provided pod must have at least one container")
> > +	}
> > +	if len(o.Attach.ContainerName) == 0 {
> > +		return fmt.Errorf("you must provide a container name to debug")
> > +	}
> > +	if containerForName(o.Attach.Pod, o.Attach.ContainerName) == nil {
> > +		return fmt.Errorf("the container %q is not a valid container name; must be one of %v", o.Attach.ContainerName, names)
> > +	}
> > +	return nil
> > +}
> > +
> > +type WatchConditionFunc func(event watch.Event) (bool, error)
> > +
> > +func Until(timeout time.Duration, watch watch.Interface, conditions ...WatchConditionFunc) error {
>
> Most of the time I've seen this attempted, people want the resulting
> runtime.Object that they were waiting for. If we agree that's the far
> more common that waiting for deletion, then returning runtime.Object,
> error makes sense. In fact, it might make sense regardless.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/7294/files#r53185188>.
>